xdsl/transforms/common_subexpression_elimination.py

```diff
@@ -250,8 +247,11 @@ def simplify(self, thing: Operation | Block | Region):
         self._commit_erasures()
 
 
-def cse(thing: Operation | Block | Region):
-    CSEDriver().simplify(thing)
+def cse(thing: Operation | Block | Region, rewriter: PatternRewriter | None = None):
```

Comment:
So the thing is, some people are using this pass inside the pattern rewriter.
Without passing the pattern rewriter here, we get some bugs because some operations are removed without notifying the pattern rewriter, meaning some operations are still in the worklist, even though they are now deleted.

So besides allowing to pass a pattern rewriter, I'm not sure what we should do here?

Reply:
I was ok with passing a pattern rewriter in, I just meant the type of the argument could be `PatternRewriter | Rewriter | None`

Reply:
So the problem here is that `dce` was called inside a region of an operation that was matched.
Their goal I think was to only optimize this sub-region, not the entire program, if that makes sense.
I think in general, I'm in favor of moving towards APIs that explicitely ask for Rewriters whenever they modify something. That should allow us to better keep track of what is modified during rewrites for instance, without having to re-rewrite the entire thing?

Reply:
For the first part of the message, the idea is that the call to `cse` in xdsl/transforms/control_flow_hoist.py is done inside a single pattern rewrite, not in the entire file. Doing the `dce` approach in the pattern rewriter would instead call `dce` in the entire program, instead of just this region.

And the relationship should be that `PatternRewriter` subclass `Rewriter`, and that's more or less my next priority.

Reply:
oh this wasn't what I was suggesting at all. I was suggesting that `cse` could take a `PatternRewriterListener` instead of a `PatternRewriter` like `dce` does but I agree now that it would be cleaner to just force it to do all its modifications through a rewriter.

Reply:
I would say leave it, and update `dce` to use a rewriter in a following PR

Reply:
I think the dce implementation is flawed anyway, as it erases a block without notifying that the containing operations are deleted

Reply:
Oh I see, then yeah I'll update `dce` in the next PR.
I realize we might also be missing some operations to erase blocks, I should fix that as well.